// sequelize-typescript-migration/dist/example/practice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const sequelize_typescript_migration_1 = require("../../sequelize-typescript-migration");
const sequelize_typescript_1 = require("sequelize-typescript");
const path = require("path");
const car_model_1 = require("models/car.model");
const car_brand_model_1 = require("models/car_brand.model");
const bootstrap = async () => {
    const sequelize = new sequelize_typescript_1.Sequelize({
        username: "kimjbstar",
        password: "12091457",
        database: "test_migration2",
        host: "localhost",
        dialect: "mysql",
        models: [car_brand_model_1.CarBrand, car_model_1.Car],
        timezone: "+09:00",
        logging: false,
    });
    try {
        const result = await sequelize_typescript_migration_1.SequelizeTypescriptMigration.makeMigration(sequelize, {
            outDir: path.join(__dirname, "./migrations"),
        });
        console.log(result);
    }
    catch (e) {
        console.log(e);
    }
};
bootstrap();
